File: code/python/3443.py
# https://leetcode.com/problems/maximum-manhattan-distance-after-k-changes/
class Solution:
    def maxDistance(self, s: str, k: int) -> int:
        shifts = {'N': (0, 1), 'S': (0, -1), 'E': (1, 0), 'W': (-1, 0)}
        x = y = 0
        dist_total = dist_prev = res = 0
        for c in s:
            x += shifts[c][0]
            y += shifts[c][1]
            dist_cur = abs(x) + abs(y)
            diff = dist_cur - dist_prev
            dist_prev = dist_cur
            if diff == -1 and k > 0:
                diff = 1
                k -= 1
            dist_total += diff
            if dist_total > res: res = dist_total
        return res

File: code/java/3443.java
// https://leetcode.com/problems/maximum-manhattan-distance-after-k-changes/
class Solution {
    public int maxDistance(String s, int k) {
        final int n = s.length();
        int res = 0, x = 0, y = 0;
        for (int i = 0; i < n; ++i) {
            char ch = s.charAt(i);
            switch (ch) {
                case 'E':
                    x++; break;
                case 'W':
                    x--; break;
                case 'N':
                    y++; break;
                case 'S':
                    y--; break;
            }
            res = Math.max(res, Math.min(Math.abs(x) + Math.abs(y) + 2 * k, i + 1));
        }
        return res;
    }
}
